Field sensitivity improvements round 2

smowton opened the pull request:
Here's my latest batch of suggested improvements. The cleanup immediately after a field-assignments is I think the main functional problem with the current draft. There are two problems I wanted to address but not quite sure how to yet:

1. `rewrite_with_to_field_symbols` reverses a transformation made by `symex_assign_rec` with a member expression, and reverses it in a way that I expect would break with field-sensitivity switched off. I suggest this particular `with -> member` transformation should happen within `symex_assign_rec`, and the direction (with -> member or member -> with) should be controlled by whether field sensitivity is switched on. The same might also apply to `shift_indexed_access_to_lhs`.

2. I still don't have a good answer for the various `apply` sites, though I'm a bit less worried now I've managed to eliminate one from `goto_symex_statet`. The remaining ones all cluster around symex_dereference and simplify.

smowton committed:
Clean up composite symbols immediately after they are assigned

This avoids having two different versions of a symbol (the field-of-composite and the indivisible field versions)
in the most direct way possible.

smowton committed:
Make is_divisible test cheaper

I appreciate synchronisation with get_fields is important, but it seems a bit expensive and isn't
hard to express as a pure test.

smowton committed:
Remove constant propagator lookup after applying field sensitivity

This should no longer be necessary now that field_sensitivity::apply uses state.rename to apply
L2 renaming, which itself consults the constant propagator.

smowton commented:
One significant concern: total test-suite execution time is 3308 seconds, up from 2320 on current develop. Next week I intend to find a few particular tests that have slowed down a lot and see if I can do anything about that.

smowton commented:
Looks like the measured slowdown was due to concurrent activity on the test machine

tautschnig commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -63,11 +63,10 @@ class field_sensitivityt
   /// \return False, if and only if, \p expr would be a single field-sensitive
   /// SSA expression.
-  bool is_divisible(goto_symex_statet &state, const ssa_exprt &expr) const;
+  bool is_divisible(const ssa_exprt &expr) const;
```
I'd suggest to thus make it `static` again. Doing so will also make it unnecessary to pass around `field_sensitivity` in a few cases.

tautschnig commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -385,26 +385,7 @@ void goto_symext::symex_assign_symbol(
-      underlying_composite, ns, field_sensitivity, goto_symex_statet::L1);
-    state.propagation.erase(underlying_composite.get_l1_object_identifier());
-  }
+  field_sensitivity.handle_assignment(state, lhs_mod, ns);
```
Maybe `post_process_assignment`?

smowton committed:
Reinstate symex-deref's `write` parameter

field_sensitivity::apply needs to avoid expanding struct-typed symbols into fields when they're
being used as lvalues.

tautschnig commented:
> 1. `rewrite_with_to_field_symbols` reverses a transformation made by `symex_assign_rec` with a member expression, and reverses it in a way that I expect would break with field-sensitivity switched off. I suggest this particular `with -> member` transformation should happen within `symex_assign_rec`, and the direction (with -> member or member -> with) should be controlled by whether field sensitivity is switched on. The same might also apply to `shift_indexed_access_to_lhs`.

There's a catch here: all these rewrites back into field-sensitive names will often only be possible after we've done our L2 renaming, and constant propagation in particular. If we were to have some "rename-to-L2-all-read-only-fields-in-LHS" then we could likely avoid that dance. But we don't currently have that...

tautschnig commented:
> 2\. I still don't have a good answer for the various `apply` sites, though I'm a bit less worried now I've managed to eliminate one from `goto_symex_statet`. The remaining ones all cluster around symex_dereference and simplify.

I think this might get resolved over time with Romain's proposal of wrapping renamed objects into tag-typed ones (`level2t<exprt>`). We could then add another tag type or maybe otherwise already know that we're safe.

smowton committed:
Comment on and clarify field-sensitivity-specific expression transformations

The function name was previously incorrect (it did more than it said), so now I just call both transformations
from symex_assign.
